### PR TITLE
Resource isolation, adding a separate scan_worker to schedule io task…

### PR DESCRIPTION
**Pull request by laotan332**
… execution

## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

### CODE REVIEW

**Review by satanson — DISMISSED** [review]

**Review by satanson — DISMISSED** [review]

File: be/src/common/config.h [hunk 1]
```diff
@@ -656,7 +656,15 @@ CONF_Int64(pipeline_sink_buffer_size, "64");
...
 
+// The max number of io tasks for each scan operator.
 CONF_Int64(pipeline_scan_max_tasks_per_operator, "4");
+// yield scan io task when maximum time in nano-seconds has spent in current execution round,
+// if it runs in the dispatcher owned by other workgroup, which has running drivers.
+CONF_Int64(pipeline_scan_task_yield_max_tims_spent, "100000000");
+// The max schedule period for adjusting io weight of each workgroup.
+CONF_Int64(pipeline_scan_task_yield_preempt_max_time_spent, "20000000");
+// The max schedule period for adjusting io weight of each workgroup.
+CONF_Int32(pipeline_max_io_schedule_num_period, "512");
```
[the remark this answers is not in the document]
Comment (laotan332) [reply be/src/common/config.h:667]: This will be modified by a separate pr

File: be/src/exec/pipeline/chunk_source.h [hunk 2]
```diff
@@ -11,6 +11,12 @@
 
 namespace starrocks {
 class RuntimeState;
+
+namespace workgroup {
+class WorkGroup;
+using WorkGroupPtr = std::shared_ptr<WorkGroup>;
+} // namespace workgroup
+
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/chunk_source.h:19]: There is a new header file `exec/workgroup/work_group_fwd.h` contains these forward declarations, so maybe we could just include this file.

File: be/src/exec/pipeline/olap_chunk_source.h [hunk 3]
```diff
@@ -17,10 +17,18 @@
...
+
 namespace vectorized {
 class RuntimeFilterProbeCollector;
 }
+
+namespace workgroup {
+class WorkGroup;
+using WorkGroupPtr = std::shared_ptr<WorkGroup>;
+} // namespace workgroup
+
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/olap_chunk_source.h:31]: Similar.

File: be/src/runtime/exec_env.h [hunk 4]
```diff
@@ -67,6 +67,10 @@ namespace pipeline {
 class DriverDispatcher;
 }
 
+namespace workgroup {
+class ScanWorker;
```
Comment (ZiheLiu) [inline be/src/runtime/exec_env.h:71]: Maybe we could add this to `exec/workgroup/work_group_fwd.h`.

**Comment by laotan332** [comment]
run starrocks_be_unittest

**Comment by wangruin** [comment]
run starrocks_be_unittest

**Review by ZiheLiu — APPROVED** [review]

**Review by satanson — APPROVED** [review]

File: be/src/exec/pipeline/scan_operator.cpp [hunk 5]
```diff
@@ -115,6 +121,9 @@ void ScanOperator::set_finishing(RuntimeState* state) {
 
 StatusOr<vectorized::ChunkPtr> ScanOperator::pull_chunk(RuntimeState* state) {
     RETURN_IF_ERROR(_try_to_trigger_next_scan(state));
+    if (_workgroup != nullptr) {
+        _workgroup->decrease_chunk_num(1);
+    }
```
Comment (liuyehcf) [inline be/src/exec/pipeline/scan_operator.cpp:126]: pull_chunk may return nullptr, is it ok to call `decrease_chunk_num` in this situation?

**Review by Pslydhh — APPROVED** [review]

File: be/src/exec/pipeline/scan_operator.cpp (hunk 5 above)
Comment (laotan332) [reply be/src/exec/pipeline/scan_operator.cpp:126]: Yes, this shouldn't really be called decrease_chunk_num, it's meant to be a request to get chunk， maybe it should rename to ask_for_chunk ?, i would rename in next pr

**Review by liuyehcf — APPROVED** [review]

File: be/src/exec/pipeline/scan_operator.cpp (hunk 5 above)
Comment (laotan332) [reply be/src/exec/pipeline/scan_operator.cpp:126]: it just try to get, but not necessarily successful， but we need count the number of it

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**